tools/codegen/gen_vmap_plumbing2.py

```diff
@@ -0,0 +1,213 @@
+from tools.codegen.api.types import (
+    DispatcherSignature,
+)
+from tools.codegen.model import (
+    BaseTy, Variant, OptionalType, BaseType, ListType, NativeFunction, Type,
+    Argument, Return, SchemaKind, Tag
+)
+from tools.codegen.context import method_with_native_function
+from tools.codegen.utils import mapMaybe
+from dataclasses import dataclass
+from typing import List, Dict, Optional, Tuple, Set, Any, Union, Sequence, TypeVar
+import textwrap
+
+
+def is_tensor(typ: Type) -> bool:
+    return isinstance(typ, BaseType) and typ.name == BaseTy.Tensor
+
+
+def is_optional_tensor(typ: Type) -> bool:
+    return isinstance(typ, OptionalType) and is_tensor(typ.elem)
+
+
+def is_vector_tensor(typ: Type) -> bool:
+    return isinstance(typ, ListType) and is_tensor(typ.elem)
+
+
+def unwrap_tensor(name: str) -> List[str]:
+    result = f"""\
+    Tensor {name}_value;
+    optional<int64_t> {name}_bdim;
+    std::tie({name}_value, {name}_bdim) = unwrapTensorAtLevel({name}, cur_level);"""
+    return textwrap.dedent(result).split('\n')
+
+
+def unwrap_optional_tensor(name: str) -> List[str]:
+    result = f"""\
+    optional<Tensor> {name}_value;
+    optional<int64_t> {name}_bdim;
+    if ({name}) {{
+        std::tie({name}_value, {name}_bdim) = unwrapTensorAtLevel({name}.value(), cur_level);
+    }}"""
+    return textwrap.dedent(result).split('\n')
+
+
+def gen_unwraps(flat_arguments: List[Argument]) -> Tuple[List[str], List[str]]:
+    arg_names = [a.name for a in flat_arguments]
+    arg_types = [a.type for a in flat_arguments]
+
+    tensors = [name for typ, name in zip(arg_types, arg_names) if is_tensor(typ)]
+    optional_tensors = [name for typ, name in zip(arg_types, arg_names) if is_optional_tensor(typ)]
+
+    unwraps = []
+    for tensor in tensors:
+        unwraps += unwrap_tensor(tensor)
+
+    for opt_tensor in optional_tensors:
+        unwraps += unwrap_optional_tensor(opt_tensor)
+    unwraps = '\n'.join(unwraps)
+
+    unwrapped_arg_list = []
+    for arg in arg_names:
+        if arg in tensors or arg in optional_tensors:
+            unwrapped_arg_list += [f'{arg}_value', f'{arg}_bdim']
+        else:
+            unwrapped_arg_list.append(arg)
+    return unwraps, unwrapped_arg_list
+
+
+def get_aten_op_call(schema) -> str:
+    if schema.name.overload_name:
+        return f'ATEN_FN2({schema.name.name}, {schema.name.overload_name})'
+    return f'ATEN_FN({schema.name.name})'
+
+
+def gen_case_where_all_bdims_are_none(flat_args: List[Argument], schema) -> str:
+    conditions = []
+    for arg in flat_args:
+        if not arg.type.is_tensor_like():
+            continue
+        conditions.append(f'!isBatchedAtLevel({arg.name}, cur_level)')
+    aten_op = get_aten_op_call(schema)
+    arg_names = [a.name for a in flat_args]
+
+    return f"""\
+if ({' && '.join(conditions)}) {{
+  return {aten_op}({', '.join(arg_names)});
```

Comment:
it looks like this calls into `at::_ops` - I think it's probably fine to refer to that directly in the codegen instead of using the `ATEN` macro, since do that internally in a bunch of other places too.

You can see an example of some codegen that calls it [here](https://github.com/pytorch/pytorch/blob/6643522db9ff595f564b8081de58b3a33c546178/tools/codegen/gen.py#L402).

One thing that Ed convinced me to do more in the codegen to help keep the API's straight in my head is use `translate()` and various `*Signature` objects more often to codegen an API call.

Basically, `at::_ops` follows the "Dispatcher API", which (documented in `dispatcher.py` has a few characteristics)
- arguments are not defaulted
- TensorOptions are unpacked into their individual arguments

And passing in `[a.name for a in flat_args]` to that function happens to work, because you generated `flat_args` from `f.func.flat_all`, which also unpacks TensorOptions.

Instead of worrying about all of that though, you can just use `translate` + `DispatcherSignature`:
```
        // ok, the translate is a bit overkill here since both the target and our current function follow
        // the dispatcher convention (but we're still making that explicit here)
        // target = Dispatcher API (because we're calling at::_ops, which follows the Dispatcher API)
        // current context = Dispatcher API (because we're codegen'ing a snippet that goes in a function that we register to the dispatcher)
        sig = DispatcherSignature.from_schema(schema)
        return f"""
if ({' && '.join(conditions)}) {{
    return at::_ops::{f.func.name.unambiguous_name()}::call({', 'join(e.expr for e in translate(sig.arguments(), sig.arguments())});
}}
"""
```

Reply:
also, totally a minor nit that you can take or leave - I find the codegen a lot easier to understand when more of the emitted code is grouped together. I think this function is small enough that you could pretty easily fold it back into the calling function(s), and merge this into the giant emitted f-string in the calling function. I guess the downside is that you need to inline this function in both places.

Reply:
I'll keep it in two different places so that we don't have to duplicate the logic and inline into both places